### PR TITLE
Can create streams on h2 connections

### PR DESCRIPTION
**Pull request by graebm**
Can create and clean up streams on h2 connections. We're not sending or receiving any data yet, just slotting things into datastructures.

Also added shutdown machinery to the h2 connection.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/h2_connection.c [hunk 1]
```diff
@@ -217,6 +423,11 @@ static int s_handler_process_read_message(
     (void)handler;
     (void)slot;
     (void)message;
+
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_connection.c]: Yes, this is purposefully unimplemented till I start hooking up decoding

File: source/h2_connection.c [hunk 2]
```diff
@@ -177,36 +253,166 @@ static void s_handler_destroy(struct aws_channel_handler *handler) {
...
+        CONNECTION_LOG(TRACE, connection, "Scheduling cross-thread work task");
+        aws_channel_schedule_task_now(connection->base.channel_slot->channel, &connection->cross_thread_work_task);
+    }
+
+    AWS_H2_STREAM_LOG(DEBUG, stream, "Created HTTP/2 request stream"); /* #TODO: print method & path */
+    return &stream->base;
+
+error:
+    /* Force destruction of the stream, avoiding ref counting */
+    stream->base.vtable->destroy(&stream->base);
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_connection.c]: I acknowledge this is weird.

The public API is `aws_http_stream_release()`, but before it invokes the the stream's v-table it does some stuff where it's assuming the stream was successfully created.

File: source/h2_connection.c [hunk 3]
```diff
@@ -177,36 +253,166 @@ static void s_handler_destroy(struct aws_channel_handler *handler) {
...
+        struct aws_h2_stream *stream = AWS_CONTAINER_OF(node, struct aws_h2_stream, node);
+        s_activate_stream(connection, stream);
+    }
+
+    /* #TODO: process stuff from other API calls (ex: window-updates) */
+}
+
+/* Set ID for new stream. Lock must be held while calling this. */
+static int s_acquire_next_stream_id(struct aws_h2_connection *connection, uint32_t *out_stream_id) {
+    if (AWS_UNLIKELY(connection->synced_data.next_stream_id > MAX_STREAM_ID)) {
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_connection.c]: yes, next_stream_id is a uint32_t

[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_connection.c]: Oops, but this code isn't totally correct
fixing

[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_connection.c]: nevermind. it's fine. the code is fine.

File: source/h2_connection.c [hunk 4]
```diff
@@ -177,36 +253,166 @@ static void s_handler_destroy(struct aws_channel_handler *handler) {
...
+    struct aws_linked_list pending_streams;
+    aws_linked_list_init(&pending_streams);
 
     { /* BEGIN CRITICAL SECTION */
-        int err = aws_mutex_lock(&connection->synced_data.lock);
-        AWS_FATAL_ASSERT(err == AWS_OP_SUCCESS);
+        s_lock_synced_data(connection);
+        connection->synced_data.is_cross_thread_work_task_scheduled = false;
+
+        aws_linked_list_swap_contents(&connection->synced_data.pending_stream_list, &pending_streams);
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_connection.c]: We'd have to totally redesign the linked list implementation to make that work, and it wouldn't be double-linked anytmore

File: source/h2_connection.c [hunk 5]
```diff
@@ -177,36 +253,166 @@ static void s_handler_destroy(struct aws_channel_handler *handler) {
...
+    return;
+error:
+    s_stream_complete(stream, aws_last_error());
+}
+
+/* Perform on-thread work that is triggered by calls to the connection/stream API */
+static void s_cross_thread_work_task(struct aws_channel_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+    if (status != AWS_TASK_STATUS_RUN_READY) {
+        return;
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_connection.c]: The task scheduler is already going to log "Running HTTP/2 cross-thread work task with <Canceled> status"

File: source/h2_connection.c [hunk 6]
```diff
@@ -79,6 +86,67 @@ static const struct aws_h2_decoder_vtable s_h2_decoder_vtable = {
     .on_data = NULL,
 };
 
+static void s_lock_synced_data(struct aws_h2_connection *connection) {
+    int err = aws_mutex_lock(&connection->synced_data.lock);
+    AWS_ASSERT(!err && "lock failed");
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_connection.c]: sticking with ASSERT for checking my internal logic
most code doesn't even bother checking lock()'s return value

File: source/h2_connection.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (JonathanHenson) [reply source/h2_connection.c]: k

**Pull request merged**